internal/client/consensus/grandpa/change_tree.go

```diff
@@ -182,59 +182,80 @@ func (ct *ChangeTree[H, N, ID]) FinalizesAnyWithDescendentIf(
 }
 
 // FinalizationResult Result of finalising a node (that could be a part of the roots or not).
-type FinalizationResult scale.VaryingDataType
+type FinalizationResult[H comparable, N constraints.Unsigned] struct {
+	inner any
+}
 
-// Set will set a VaryingDataTypeValue using the underlying VaryingDataType
-func (fr *FinalizationResult) Set(val scale.VaryingDataTypeValue) (err error) {
-	vdt := scale.VaryingDataType(*fr)
-	err = vdt.Set(val)
-	if err != nil {
+type FinalizationResultValues[H comparable, N constraints.Unsigned] interface {
+	changed[H, N] | unchanged
+}
+
+func setFinalizationResult[
+	H comparable, N constraints.Unsigned, Value FinalizationResultValues[H, N],
+](mvdt *FinalizationResult[H, N], value Value) {
+	mvdt.inner = value
+}
+
+func (mvdt *FinalizationResult[H, N]) SetValue(value any) (err error) {
+	switch value := value.(type) {
+	case changed[H, N]:
+		setFinalizationResult[H, N](mvdt, value)
 		return
+	case unchanged:
+		setFinalizationResult[H, N](mvdt, value)
+		return
+	default:
+		return fmt.Errorf("unsupported type")
```

Comment:
I would to suggest panicking here. That is the tradeoff of using `any + constraint types`. The type is evaluated on runtime, which is something we can avoid using normal interfaces to achieve a VDT like style with comptime evaluation

Reply:
This is to satisfy the `VaryingDataType` interface, where the method `SetValue` can return an error.  This is used in the scale codec to decode into.  I'd consider revising the `VaryingDataType` interface if we were able to do this without returning an error using your suggestion.